README.md
<div align="center">
  <h1>
    <br />
    <div>:tada:</div>
    <br />
    <div>hyper-unlimited-power</div>
    <br />
  </h1>
  <br />
  <div>A configurable <a href="https://hyper.is/">hyper</a> power-mode plugin</div>
  <br />
  <a href="https://www.npmjs.com/package/hyper-unlimited-power"><img src="https://img.shields.io/npm/v/hyper-unlimited-power?style=flat-square&color=blue" /></a>
  <a href="https://www.npmjs.com/package/hyper-unlimited-power"><img src="https://img.shields.io/npm/dm/hyper-unlimited-power?style=flat-square&color=blue" /></a>
  <a href="https://hyper.is/"><img src="https://img.shields.io/badge/plugin-%E2%9A%A1%EF%B8%8Fhyper-blue?style=flat-square" /></a>
  <br /><br />
  <pre>hyper i <a href="https://www.npmjs.com/package/hyper-unlimited-power">hyper-unlimited-power</a></pre>
  <h1></h1>
</div>

<p align="center">
  <img src="https://user-images.githubusercontent.com/9451626/111384154-14196f80-86a1-11eb-80a6-e45583f61917.gif" />
</p>

## Installation

Add the following to your `~/.hyper.js` plugin array.

```javascript
module.exports = {
  ...
  plugins: [
    'hyper-unlimited-power'
  ]
  ...
};
```

## Config Options

Add any of the following to your `~/.hyper.js` config object.

### Enable Rainbow Particle Colors

<p align="center">
  <img src="https://user-images.githubusercontent.com/9451626/111384544-8d18c700-86a1-11eb-8f89-cc3c8fd1ad57.gif" />
</p>

Type: `bool`\
Default value: `true`

```javascript
module.exports = {
  config: {
    ...
    hyperUnlimitedPower: {
      isRainbowEnabled: true
    }
    ...
  }
};
```

### Enable Combo Mode

<p align="center">
  <img src="https://user-images.githubusercontent.com/9451626/111672467-e7d13080-8811-11eb-8653-d1f9d01f95b8.gif" />
</p>

Type: `bool`\
Default value: `true`

```javascript
module.exports = {
  config: {
    ...
    hyperUnlimitedPower: {
      isComboEnabled: true
    }
    ...
  }
};
```

### Enable Permanent Screen Shake

<p align="center">
  <img src="https://user-images.githubusercontent.com/9451626/111384472-74101600-86a1-11eb-895a-05f868b052c7.gif" />
</p>

Type: `bool`\
Default value: `false`

```javascript
module.exports = {
  config: {
    ...
    hyperUnlimitedPower: {
      isEarthquakeEnabled: true
    }
    ...
  }
};
```

### Static Particle Colors

<p align="center">
  <img src="https://user-images.githubusercontent.com/9451626/111384576-96a22f00-86a1-11eb-9a5e-83f304a2f9dc.gif" />
</p>

Type: `"false"|string|string[]`\
Default value: `false`

```javascript
module.exports = {
  config: {
    ...
    hyperUnlimitedPower: {
      isRainbowEnabled: false,
      staticParticleColors: ['red', 'yellow']
    }
    ...
  }
};
```

### Maximum Number of Particles

<p align="center">
  <img src="https://user-images.githubusercontent.com/9451626/111384710-be919280-86a1-11eb-9e5c-13a46396393b.gif" />
</p>

Type: `number`\
Default value: `20`

```javascript
module.exports = {
  config: {
    ...
    hyperUnlimitedPower: {
      maxParticleCount: 50
    }
    ...
  }
};
```

### Particle Velocity Range

<p align="center">
  <img src="https://user-images.githubusercontent.com/9451626/111384762-c7826400-86a1-11eb-92ec-01ad603798f3.gif" />
</p>

Type: `object`\
Default value: `{ x: [-1, 1], y: [-2, -0.5] }`

```javascript
module.exports = {
  config: {
    ...
    hyperUnlimitedPower: {
      particleVelocityRange: {
        x: [-1, 1],
        y: [-2.5, -1.5]
      }
    }
    ...
  }
};
```

## Secret Config Options

#### `comboActivationThreshold`

Type: `number`\
Default value: `15`

#### `comboTimeUntilDecay`

Type: `number`\
Default value: `2000`

#### `particleAlphaFadeout`

Type: `number`\
Default value: `0.92`

#### `particleAlphaMinThreshold`

Type: `number`\
Default value: `0.1`

#### `particleGravity`

Type: `number`\
Default value: `0.075`

#### `shakeIntensity`

Type: `number`\
Default value: `1`

## License

MIT - see the [LICENSE.md](https://github.com/eels/hyper-unlimited-power/blob/main/LICENSE.md) file for details

## Acknowledgments & Credits

* Inspired by the original [hyperpower](https://github.com/vercel/hyperpower) plugin
* Which in turn was based on [power-mode](https://atom.io/packages/power-mode) and [RagePower](https://github.com/itszero/rage-power)
* With additional fixes from [MajorBreakfast](https://github.com/MajorBreakfast) in ([#75](https://github.com/vercel/hyperpower/pull/75))


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v1.0.4](https://github.com/eels/hyper-unlimited-power/compare/v1.0.3...v1.0.4) - 2022-01-17

### Commits

- build(deps): bump actions/setup-node from 2.5.0 to 2.5.1 [`ad571c9`](https://github.com/eels/hyper-unlimited-power/commit/ad571c9c5e7b19dcd63af93e9dafabc7429f255a)
- chore: add bug issue template [`0ce94ad`](https://github.com/eels/hyper-unlimited-power/commit/0ce94ad532ba844f28d444118ab93cc3b59e15cc)
- chore: add code of conduct [`9717cde`](https://github.com/eels/hyper-unlimited-power/commit/9717cde04b23d73297574da177d851cc2032a777)
- chore: add contributing.md [`f5e773d`](https://github.com/eels/hyper-unlimited-power/commit/f5e773d77ba365cfcee923358700f911c64003a6)
- chore: add pull request template [`990b587`](https://github.com/eels/hyper-unlimited-power/commit/990b587d185de93c82ebf5a27eb6e97319f341e6)
- chore(deps-dev): bump @commitlint/cli from 15.0.0 to 16.0.1 [`96f451f`](https://github.com/eels/hyper-unlimited-power/commit/96f451fef0d4d3ef1d88d6c7abfa9e58da94b52b)
- chore(deps-dev): bump @commitlint/config-conventional [`ba953f7`](https://github.com/eels/hyper-unlimited-power/commit/ba953f749b8b8545a88b36915cd06d24bff4291a)
- ci: add dependabot auto merge workflow [`7f72017`](https://github.com/eels/hyper-unlimited-power/commit/7f720179da96cb4aa2fe228b3a5b1b0b63b3d508)
- docs: update v1.0.3 changelog [`61d4e58`](https://github.com/eels/hyper-unlimited-power/commit/61d4e5878084fcb69dbe5ddb495a7dbae2f07321)
- fix: ensure next particle state is returned in draw frame [`a58c068`](https://github.com/eels/hyper-unlimited-power/commit/a58c06823f94a3ca57628e465eaecb0b549e30ee)

## [v1.0.3](https://github.com/eels/hyper-unlimited-power/compare/v1.0.2...v1.0.3) - 2021-12-21

### Commits

- Add badges to README.md [`6b3a3a5`](https://github.com/eels/hyper-unlimited-power/commit/6b3a3a54f9896970abd7421e8557ab3cd75726de)
- Add husky + setup [`8da3160`](https://github.com/eels/hyper-unlimited-power/commit/8da316064b5d82dc18ab65ba3c2a087580862764)
- Add lint-staged and add explicit extensions to some dot files [`e2fad5e`](https://github.com/eels/hyper-unlimited-power/commit/e2fad5e3f0d59e50d55cedec0833171187645c25)
- build: add dependabot.yml [`d659514`](https://github.com/eels/hyper-unlimited-power/commit/d6595149e5b17602a28ac544a0a055efd82c17c5)
- build: add github action workflows [`d1501d9`](https://github.com/eels/hyper-unlimited-power/commit/d1501d96e80e7084b5d71dc4bb485dfefa49ee76)
- build(deps): bump actions/setup-node from 1 to 2.4.1 [`be16981`](https://github.com/eels/hyper-unlimited-power/commit/be16981e10b39d76056cc001803a9246ea3b643d)
- build(deps): bump actions/setup-node from 2.4.1 to 2.5.0 [`ab6722c`](https://github.com/eels/hyper-unlimited-power/commit/ab6722cc890808dd7d00401e58e1fc7849718222)
- chore: add .prettierignore [`3049e6e`](https://github.com/eels/hyper-unlimited-power/commit/3049e6e256bdcc0fa082f8662960ac84ec1fc469)
- chore: add vscode directory with sensible defaults [`a3ecc66`](https://github.com/eels/hyper-unlimited-power/commit/a3ecc6650dba71deeb7bbb87951d4985ece4039c)
- chore: update dependency versions [`b489bce`](https://github.com/eels/hyper-unlimited-power/commit/b489bce52b37505d931cf003281cc88d52322e80)

## [v1.0.2](https://github.com/eels/hyper-unlimited-power/compare/v1.0.1...v1.0.2) - 2021-03-28

### Commits

- Fix incorrect dependency inclusion [`33f7160`](https://github.com/eels/hyper-unlimited-power/commit/33f71604f6a995bdd05045a9c66af2378e1fefe9)

## [v1.0.1](https://github.com/eels/hyper-unlimited-power/compare/v1.0.0...v1.0.1) - 2021-03-28

### Commits

- Unscope project due to Hyper issues and update README.md [`971cf11`](https://github.com/eels/hyper-unlimited-power/commit/971cf11e542c54b861b6c36784384a6779b6cd06)

## v1.0.0 - 2021-03-28

### Commits

- Add access-level to package.json [`4f3f4b3`](https://github.com/eels/hyper-unlimited-power/commit/4f3f4b399c0047dca2533203db7cc40ddbcf3d2e)
- Change package.json in preparation to publish [`fc8dcab`](https://github.com/eels/hyper-unlimited-power/commit/fc8dcab50a5e29cfe18c0a4573d4d9cf48690106)
- Fix broken README.md formatting [`395f4fa`](https://github.com/eels/hyper-unlimited-power/commit/395f4faab739435cc2b450526da67b4ae26cd91d)
- Fix broken README.md formatting [`7455458`](https://github.com/eels/hyper-unlimited-power/commit/7455458fcb3bd9272f1bc88d02a8aa66d8de9cef)
- Initial commit [`9676fdb`](https://github.com/eels/hyper-unlimited-power/commit/9676fdb9855addfd6c9f797da461c700f829dcb9)
- Roll back version number for initial release [`4eccc0c`](https://github.com/eels/hyper-unlimited-power/commit/4eccc0c052edac8e8ca7319616bbd41737366fcb)


.eslintrc.js
module.exports = {
  "extends": [
    "eslint:recommended",
    "standard",
    "semistandard",
    "standard-react",
  ],

  "plugins": [
    "react",
    "sort-destructure-keys",
    "sort-keys-fix"
  ],

  "settings": {
    "react": {
      "version": "latest"
    }
  },

  "rules": {
    "array-bracket-spacing": [
      "error",
      "never"
    ],
    "arrow-parens": [
      "error",
      "always"
    ],
    "newline-before-return": [
      "error"
    ],
    "newline-after-var": [
      "error",
      "always"
    ],
    "no-console": [
      "error"
    ],
    "no-multiple-empty-lines": [
      "error",
      {
        "max": 1,
        "maxEOF": 1,
        "maxBOF": 0
      }
    ],
    "no-undef": [
      "error"
    ],
    "one-var": [
      "error",
      "never"
    ],
    "padding-line-between-statements": [
      "error",
      {
        "blankLine": "never",
        "next": "import",
        "prev": "import"
      }
    ],
    "space-before-function-paren": [
      "error",
      "never"
    ],
    "quotes": [
      "error",
      "single"
    ],
    "quote-props": [
      "error",
      "consistent"
    ],
    "yoda": [
      "error",
      "always"
    ],

    "react/sort-comp": [
      "error",
      {
        "order": [
          "static-methods",
          "lifecycle",
          "getters",
          "setters",
          "method-handlers",
          "method-getters",
          "method-setters",
          "everything-else",
          "rendering"
        ],
        "groups": {
          "method-getters": [
            "/^get.+$/"
          ],
          "method-handlers": [
            "/^handle.+$/"
          ],
          "method-setters": [
            "/^set.+$/"
          ],
          "rendering": [
            "/^render.+$/",
            "render"
          ],
        }
      }
    ],
    "react/sort-prop-types": [
      "error",
      {
        "ignoreCase": false,
        "callbacksLast": false,
        "noSortAlphabetically": false,
        "requiredFirst": false,
        "sortShapeProp": true
      }
    ],

    "sort-destructure-keys/sort-destructure-keys": [
      "error",
      {
        "caseSensitive": true
      }
    ],

    "sort-keys-fix/sort-keys-fix": [
      "error"
    ]
  }
};


src/index.js
const Color = require('color');
const PropTypes = require('prop-types');
const colorNameToHex = require('convert-css-color-name-to-hex');
const throttle = require('lodash.throttle');

exports.decorateTerm = (Term, { React }) => {
  const HYPER_UNLIMITED_CONFIG = {};
  const config = require('@electron/remote').app.config.getConfig();
  const configEntries = Object.entries(config.hyperUnlimitedPower || {});

  for (const [key, value] of configEntries) {
    HYPER_UNLIMITED_CONFIG[key] = value;
  }

  class DecoratedTerm extends React.Component {
    constructor(props, context) {
      super(props, context);

      this.comboDecay = this.comboDecay.bind(this);
      this.config = this.setConfigurationOptions();
      this.drawFrame = this.drawFrame.bind(this);
      this.getColorSpectrum = this.getColorSpectrum.bind(this);
      this.handleOnCursorMove = this.handleOnCursorMove.bind(this);
      this.handleOnDecorated = this.handleOnDecorated.bind(this);
      this.initCanvas = this.initCanvas.bind(this);
      this.resizeCanvas = this.resizeCanvas.bind(this);
      this.shake = throttle(this.shake.bind(this), 100, { trailing: false });
      this.spawnParticles = throttle(this.spawnParticles.bind(this), 25, { trailing: false });

      this.canvas = null;
      this.canvasCxt = null;
      this.comboDecayTimer = null;
      this.div = null;
      this.particles = [];
      this.state = this.initialState;
    }

    componentWillUnmount() {
      if (this.canvas !== null) {
        document.body.removeChild(this.canvas);
      }
    }

    get initialState() {
      return {
        cursorMovementsMade: 0,
        isComboActive: false,
      };
    }

    get initialConfigState() {
      return {
        comboActivationThreshold: 15,
        comboTimeUntilDecay: 2000,
        isComboEnabled: true,
        isEarthquakeEnabled: false,
        isRainbowEnabled: true,
        maxParticleCount: 20,
        particleAlphaFadeout: 0.96,
        particleAlphaMinThreshold: 0.1,
        particleGravity: 0.075,
        particleVelocityRange: {
          x: [-1, 1],
          y: [-2.5, -1.5],
        },
        shakeIntensity: 1,
        staticParticleColors: false,
      };
    }

    handleOnCursorMove(cursorFrame) {
      const { onCursorMove } = this.props;
      const { cursorMovementsMade, isComboActive } = this.state;
      const { comboActivationThreshold, comboTimeUntilDecay } = this.config;
      const { isComboEnabled, isEarthquakeEnabled } = this.config;
      const { x, y } = cursorFrame;
      const origin = this.div.getBoundingClientRect();
      const nextCursorMovementsMade = cursorMovementsMade + 1;
      const isBelowComboActivation = comboActivationThreshold < nextCursorMovementsMade;
      const nextComboActiveState = isComboEnabled && (isComboActive || isBelowComboActivation);

      if (isComboEnabled) {
        clearTimeout(this.comboDecayTimer);

        this.comboDecayTimer = setTimeout(() => this.comboDecay(), comboTimeUntilDecay);
        this.setState({
          cursorMovementsMade: nextCursorMovementsMade,
          isComboActive: nextComboActiveState,
        });
      }

      const shouldShakeCanvas = isEarthquakeEnabled || nextComboActiveState;

      onCursorMove && onCursorMove(cursorFrame);
      shouldShakeCanvas && this.shake();

      window.requestAnimationFrame(() => {
        this.spawnParticles(x + origin.left, y + origin.top);
      });
    }

    handleOnDecorated(term) {
      const { onDecorated } = this.props;

      onDecorated && onDecorated(term);

      this.div = term ? term.termRef : null;
      this.initCanvas();
    }

    getColorSpectrum() {
      const { colors, cursorColor } = this.props;
      const { isRainbowEnabled, staticParticleColors } = this.config;

      if (isRainbowEnabled) {
        return Object.values(colors).map(this.colorToHex);
      }

      if (staticParticleColors !== false) {
        return Array.isArray(staticParticleColors) ? staticParticleColors : [staticParticleColors];
      }

      return [this.colorToHex(cursorColor)];
    }

    setConfigurationOptions() {
      return Object.assign(this.initialConfigState, HYPER_UNLIMITED_CONFIG);
    }

    colorToHex(name) {
      return Color(colorNameToHex(name)).hexString();
    }

    comboDecay() {
      this.setState(this.initialState);
    }

    createParticle(x, y, color) {
      const { particleVelocityRange } = this.config;

      const getVelocityModifier = (axis) => {
        const modifier = particleVelocityRange[axis][1] - particleVelocityRange[axis][0];

        return Math.random() * modifier;
      };

      const alpha = 1;
      const velocity = {
        x: particleVelocityRange.x[0] + getVelocityModifier('x'),
        y: particleVelocityRange.y[0] + getVelocityModifier('y'),
      };

      return { alpha, color, velocity, x, y };
    }

    drawFrame() {
      const { needsRedraw } = this.props;
      const { maxParticleCount } = this.config;
      const { particleAlphaFadeout, particleAlphaMinThreshold, particleGravity } = this.config;

      if (this.particles.length) {
        this.canvasCxt.clearRect(0, 0, this.canvas.width, this.canvas.height);
      }

      for (const particle of this.particles) {
        particle.velocity.y += particleGravity;
        particle.x += particle.velocity.x;
        particle.y += particle.velocity.y;
        particle.alpha *= particleAlphaFadeout;

        if (particle.alpha < particleAlphaMinThreshold) {
          continue;
        }

        this.canvasCxt.fillStyle = `rgba(${particle.color.join(',')}, ${particle.alpha})`;
        this.canvasCxt.fillRect(Math.round(particle.x - 1), Math.round(particle.y - 1), 3, 3);
      }

      const getNextParticleState = (particles) => {
        particles = [].concat(particles);
        particles = particles.slice(Math.max(particles.length - maxParticleCount, 0));
        particles = particles.filter((prt) => prt.alpha > particleAlphaMinThreshold);

        return particles;
      };

      this.particles = getNextParticleState(this.particles);

      if (this.particles.length > 0 || needsRedraw) {
        window.requestAnimationFrame(this.drawFrame);
      }

      this.props.needsRedraw = this.particles.length === 0;
    }

    generateParticleAmount() {
      return 5 + Math.round(Math.random() * 5);
    }

    initCanvas() {
      this.canvas = document.createElement('canvas');
      this.canvas.style.position = 'absolute';
      this.canvas.style.top = '0';
      this.canvas.style.pointerEvents = 'none';
      this.canvas.width = window.innerWidth;
      this.canvas.height = window.innerHeight;
      this.canvasCxt = this.canvas.getContext('2d');

      document.body.appendChild(this.canvas);
      window.requestAnimationFrame(this.drawFrame);
      window.addEventListener('resize', this.resizeCanvas);
    }

    resizeCanvas() {
      this.canvas.width = window.innerWidth;
      this.canvas.height = window.innerHeight;
    }

    shake() {
      const { shakeIntensity } = this.config;
      const intensity = shakeIntensity + 2 * Math.random();
      const x = intensity * (Math.random() > 0.5 ? -1 : 1);
      const y = intensity * (Math.random() > 0.5 ? -1 : 1);

      this.div.style.transform = `translate3d(${x}px, ${y}px, 0)`;

      setTimeout(() => {
        if (this.div) {
          this.div.style.transform = '';
        }
      }, 75);
    }

    spawnParticles(x, y) {
      const length = this.particles.length;
      const spectrum = this.getColorSpectrum();
      const particleAmount = this.generateParticleAmount();
      const particleArray = Array.from(Array(particleAmount).keys());

      for (const index of particleArray) {
        const colorCode = spectrum[index % spectrum.length];
        const r = parseInt(colorCode.slice(1, 3), 16);
        const g = parseInt(colorCode.slice(3, 5), 16);
        const b = parseInt(colorCode.slice(5, 7), 16);
        const colorArray = [r, g, b];

        this.particles.push(this.createParticle(x, y, colorArray));
      }

      if (length === 0) {
        window.requestAnimationFrame(this.drawFrame);
      }
    }

    render() {
      const propsToForward = {
        onCursorMove: this.handleOnCursorMove,
        onDecorated: this.handleOnDecorated,
      };

      return React.createElement(Term, Object.assign({}, this.props, propsToForward));
    }
  }

  DecoratedTerm.propTypes = {
    colors: PropTypes.object,
    cursorColor: PropTypes.string,
    needsRedraw: PropTypes.bool,
    onCursorMove: PropTypes.func,
    onDecorated: PropTypes.func,
  };

  return DecoratedTerm;
};
